module/zfs/zstd.c

```diff
@@ -0,0 +1,617 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2016-2018 by Klara Systems Inc.
+ * Copyright (c) 2016-2018 Allan Jude <allanjude@freebsd.org>.
+ */
```

Comment:
You're going to want to move this in to the `module/zstd/`.  What we want to do is be able to load the new the kmod independently and have it initialize and manage its own caches.   Only the compress/decompress functions required by ZFS should be exported for use.  We should also consider adding a `zfs_` prefix for at least the public functions to ensure we can't collide with the kernel provided symbols.  (they all appear to be ZSTD_*).

Reply:
since the zstd functions arent exported in my new version this will never happen. only the zstd implementation is exported